doc/src/tutorial/codegen.rst

```diff
@@ -0,0 +1,516 @@
+------------------------------
+This is where the meat of code generation is, the translation of SymPy
+expressions to specific code.Supported languages are C (``ccode``), Fortran 95 (``fcode``), 
+Javascript (``jscode``), Mathematica (``mathematica_code``), Octave/Matlab (``octave_code``), 
+Python (print_python, which is actually more like a lightweight version 
+of codegen for Python, and ``lambdarepr``, which supports many libraries 
+(like NumPy), and theano (``theano_code``).The code printers 
+are special cases of the other prints in SymPy (str printer, pretty printer, etc.).
+
+An important distinction is that the code printer has to deal with 
+assignments (using the ``sympy.printing.codeprinter.Assignment`` object).This serves
+as building blocks for the code printers and hence the ``codegen`` module.
+An example that shows the use of ``Assignment``::
+
+    >>> from sympy import symbols, MatrixSymbol, Matrix
+    >>> from sympy.printing.codeprinter import Assignment
+    >>> x, y, z = symbols('x y z')
+    >>> mat = Matrix([x, y, z]).T
+    >>> known_mat = MatrixSymbol('K', 1, 3)
+    >>> Assignment(known_mat, mat)  # doctest: +SKIP
+    Assignment(K, Matrix([[x, y, z]]))
+    >>> Assignment(known_mat, mat).lhs
+    K
+    >>> Assignment(known_mat, mat).rhs  # doctest: +SKIP
+    Matrix([[x, y, z]])
+
+Examples::
+
+    >>> from sympy import ccode, symbols, Rational
+    >>> Z, k, e, r = symbols('Z k e r')
+    >>> expr = (Rational(-1, 2)*Z*k*(e**2)/r)
+    >>> expr # doctest: +SKIP
+        2   
+    -Z⋅e ⋅k 
+    ────────
+      2⋅r   
+    >>> ccode(expr) # doctest: +SKIP
+    -1.0L/2.0L*Z*pow(e, 2)*k/r
+    >>> ccode(expr, assign_to="E") # doctest: +SKIP
+    E = -1.0L/2.0L*Z*pow(e, 2)*k/r;
+
+``Piecewise`` expressions are converted into conditionals. If an
+``assign_to`` variable is provided an if statement is created, otherwise
+the ternary operator is used. Note that if the ``Piecewise`` lacks a
+default term, represented by ``(expr, True)`` then an error will be thrown.
+This is to prevent generating an expression that may not evaluate to
+anything. A use case for ``Piecewise``::
+
+    >>> from sympy import symbols, fcode, Piecewise
+    >>> x, tau = symbols('x, tau')
+    >>> expr = Piecewise((x + 1, x > 0), (x, True))
+    >>> print(fcode(expr, tau))
+          if (x > 0) then
+             tau = x + 1
+          else
+             tau = x
+          end if
+
+The various printers also tend to support ``Indexed`` objects well.
+
+With ``contract=True`` these expressions will be turned into loops, whereas
+``contract=False`` will just print the assignment expression that should be
+looped over::
+
+    >>> from sympy import Eq, IndexedBase, Idx
+    >>> from sympy import jscode
+    >>> len_y = 5
+    >>> y = IndexedBase('y', shape=(len_y,))
+    >>> t = IndexedBase('t', shape=(len_y,))
+    >>> Dy = IndexedBase('Dy', shape=(len_y-1,))
+    >>> i = Idx('i', len_y-1)
+    >>> e = Eq(Dy[i], (y[i+1]-y[i])/(t[i+1]-t[i]))
```

Comment:
I think that spaces are recommended around operators except in slices/indexing, `[i+1]`. You can remove the spaces for operators in a string that have precedence. See the link I sent.